### ISSUE TITLE
PHP 7.2: Call to undefined function mcrypt_encrypt()

### ISSUE THREAD

**Issue by Fossil01** [nilsteampassnet/TeamPass#2094]
I've read that mcrypt wasn't required anymore in the latest version of Teampass but my webserver is still throwing:

```
2018/01/16 15:00:29 [error] 14072#14072: *4690 FastCGI sent in stderr: "PHP message: PHP Fatal error:  Uncaught Error: Call to undefined function mcrypt_encrypt() in /var/www/teampass/sources/main.functions.php:114
```

Any fix?

**Comment by Fossil01** [comment]
Never mind. I pulled master and needed to checkout 2.1.27.10

**Comment by nilsteampassnet** [comment]
Yes you are right, it is my mistake.
I need to remove this check in which this function is used as it is only for legacy users.

Thank you

**Comment by nilsteampassnet** [comment]
Reopen if needed

**Comment by thefirstben** [comment]
Hello, i am sorry i made a duplicate, i was trying to add more info.
I see the bug is closed, does this mean that it should work or that you are working on it ?
Sorry for bothering you on this, i really like this Teampass that i use on another platform, you did a great job.
I would like to understand if i should be able to make it work with PHP 7.2
From what is see the older mcrypt_encrypt/decrypt are removed in 7.2

```bash
$ grep -ns mcrypt_ *
main.functions.php:97:                mcrypt_encrypt( 
main.functions.php:102:                    mcrypt_create_iv(
main.functions.php:103:                        mcrypt_get_iv_size(MCRYPT_RIJNDAEL_256, MCRYPT_MODE_ECB),
main.functions.php:114:            mcrypt_encrypt(
main.functions.php:119:                mcrypt_create_iv(
main.functions.php:120:                    mcrypt_get_iv_size(MCRYPT_RIJNDAEL_256, MCRYPT_MODE_ECB),
main.functions.php:137:            mcrypt_decrypt(
main.functions.php:142:                mcrypt_create_iv(
main.functions.php:143:                    mcrypt_get_iv_size(MCRYPT_RIJNDAEL_256, MCRYPT_MODE_ECB),
main.functions.php:152:        mcrypt_decrypt(
main.functions.php:157:            mcrypt_create_iv(
main.functions.php:158:                mcrypt_get_iv_size(MCRYPT_RIJNDAEL_256, MCRYPT_MODE_ECB),
main.functions.php:194:    $init_vect = mcrypt_create_iv(mcrypt_get_iv_size(MCRYPT_RIJNDAEL_256, 'ctr'), MCRYPT_RAND);
main.functions.php:201:    $encrypted = mcrypt_encrypt(MCRYPT_RIJNDAEL_256, $key, $decrypted, 'ctr', $init_vect);
main.functions.php:248:    $decrypted = rtrim(mcrypt_decrypt(MCRYPT_RIJNDAEL_256, $key, $encrypted, 'ctr', $init_vect), "\0\4");
```
Line 97 -> 158 is for encryptOld & decryptOld functions
**however line 194 -> 248 are for encrypt and decrypt newer functions**

Thanks

**Comment by nilsteampassnet** [comment]
Teampass has not been tested in php 7.2.
mcrypt is a first step, but I know that each function has also being deprecated in this version.
And Teampass uses it.
So I believe that there is a complete review to do to move to 7.2

**Comment by thefirstben** [comment]
Ok thanks for the info ! 
From what i understand openssl_encrypt / openssl_encrypt seem to be the functions that replace mcrypt functions but their variables and output are different. I'll try to look into that part if i can help in some way.
Anyway, thanks for your great app

**Comment by nilsteampassnet** [comment]
Thanks.

Teampass doesn't require anymore mcrypt.
It is only in the code for legacy versions willing to upgrade to newest version.

**Comment by thefirstben** [comment]
Ok i understand.
Here is what i see if this can help you 

Using the Master Branch i get the following message in my webserver logs when trying to connect (login never finishes). This is why i was talking about mcrypt as it stops me from doing anything :
```bash
2018-02-06 23:16:03: (mod_fastcgi.c.434) FastCGI-stderr: PHP Fatal error:  Uncaught Error: Call to undefined function mcrypt_encrypt() in /usr/share/webapps/TeamPass/sources/main.functions.php:114
2018-02-06 23:16:03: (mod_fastcgi.c.434) FastCGI-stderr: Stack trace:
2018-02-06 23:16:03: (mod_fastcgi.c.434) FastCGI-stderr: #0 /usr/share/webapps/TeamPass/sources/identify.php(325): encryptOld('1234')
2018-02-06 23:16:03: (mod_fastcgi.c.434) FastCGI-stderr: #1 /usr/share/webapps/TeamPass/sources/identify.php(237): identifyUser('bQIxpCIpelqw2dH...')
2018-02-06 23:16:03: (mod_fastcgi.c.434) FastCGI-stderr: #2 {main}
2018-02-06 23:16:03: (mod_fastcgi.c.434) FastCGI-stderr:   thrown in /usr/share/webapps/TeamPass/sources/main.functions.php on line 114
```

Using the dev branch (which i guess you updated all the mycrypt stuff) i get passed the login page 
but get a red banner when i try to login : "Required PHP Extension mcrypt is not loaded", however when i use firefox console i see that there is an error in the the jquery.js (SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data):

```javascript
SyntaxError: JSON.parse: unexpected character at line 1 column 1 of the JSON data
[Learn More]
jquery.js:550:11
parseJSON
https://localhost/teampass/includes/js/jquery-ui/external/jquery/jquery.js:550:11
refreshListLastSeenItems/<
https://localhost/teampass/index.php:420:24
fire
https://localhost/teampass/includes/js/jquery-ui/external/jquery/jquery.js:3048:10
fireWith
https://localhost/teampass/includes/js/jquery-ui/external/jquery/jquery.js:3160:7
done
https://localhost/teampass/includes/js/jquery-ui/external/jquery/jquery.js:8235:5
callback
https://localhost/teampass/includes/js/jquery-ui/external/jquery/jquery.js:8778:8
```

**Comment by thefirstben** [comment]
Hello,

This was solved with version 2.1.27.11 for the installation.

However it is not solved for the upgrade process (in file : upgrade_run_defuse_for_pwds.php) : 
`[Thu May 17 11:10:17.112420 2018] [:error] [pid 1313] [client -:54018] PHP Fatal error:  Uncaught Error: Class 'PHP_Crypt\\PHP_Crypt' not found in /var/www/html/TeamPass/sources/main.functions.php:298\nStack trace:\n#0 /var/www/html/TeamPass/install/upgrade_run_defuse_for_pwds.php(113): cryption_phpCrypt('', 'none\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00...', '', 'decrypt')\n#1 {main}\n  thrown in /var/www/html/TeamPass/sources/main.functions.php on line 298, referer: https://teampass.xxxxxx/install/upgrade.php`

I have this problem when trying to upgrade to version 2.1.27.14 during step 4

**Comment by L30Bola** [comment]
I just tried updating from version 2.1.27.7 to 2.1.27.35 and got the same error as @thefirstben:

```
[Fri Apr 26 16:40:40.460819 2019] [php7:error] [pid 12] [client 172.18.0.1:39118] PHP Fatal error: Uncaught Error: Class 'PHP_Crypt\\PHP_Crypt' not found in /var/www/localhost/htdocs/teampass/sources/main.functions.php:298\nStack trace:\n#0 /var/www/localhost/htdocs/teampass/install/upgrade_run_defuse_for_pwds.php(113): cryption_phpCrypt('', 'none\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00\\x00...', '', 'decrypt')\n#1 {main}\n  thrown in /var/www/localhost/htdocs/teampass/sources/main.functions.php on line 298, referer: http://127.0.0.1/teampass/install/upgrade.php
```
